sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/evaluate/ResourceEvaluationStage.java

```diff
@@ -0,0 +1,189 @@
+
+                    offerResourcePool.release(new MesosResource(
+                            ResourceUtils.getUnreservedResource(resourceRequirement.getName(), unreserveValue)));
+                    offerRecommendationSlate.addUnreserveRecommendation(
+                            new UnreserveOfferRecommendation(offerResourcePool.getOffer(), unreserveResource));
+                    fulfilledResource = getFulfilledResource(new MesosResource(resourceRequirement.getResource()));
+                }
+
+                if (ValueUtils.compare(reserveValue, ValueUtils.getZero(reserveValue.getType())) > 0) {
+                    logger.info("Updates reserved resource with additional reservation");
+                    Resource reserveResource = ResourceUtils.getDesiredResource(
+                            resourceRequirement.getRole(),
+                            resourceRequirement.getPrincipal(),
+                            resourceRequirement.getName(),
+                            reserveValue);
+
+                    if (offerResourcePool.consume(new ResourceRequirement(reserveResource)).isPresent()) {
+                        reserveResource = ResourceUtils.setResourceId(
+                                reserveResource, resourceRequirement.getResourceId());
+                        offerRecommendationSlate.addReserveRecommendation(
+                                new ReserveOfferRecommendation(offerResourcePool.getOffer(), reserveResource));
+                        fulfilledResource = getFulfilledResource(new MesosResource(resourceRequirement.getResource()));
+                    } else {
+                        throw new OfferEvaluationException("Insufficient resources to increase resource usage.");
+                    }
+                }
+            }
+        } else {
+            if (resourceRequirement.reservesResource()) {
+                logger.info("Reserves Resource");
+                offerRecommendationSlate.addReserveRecommendation(
+                        new ReserveOfferRecommendation(offerResourcePool.getOffer(), fulfilledResource));
+            }
+
+            if (resourceRequirement.createsVolume()) {
+                logger.info("Creates Volume");
+                offerRecommendationSlate.addCreateRecommendation(
+                        new CreateOfferRecommendation(offerResourcePool.getOffer(), fulfilledResource));
+            }
+        }
+
+        logger.info("Satisfying resource requirement: {}\nwith resource: {}",
+                TextFormat.shortDebugString(resourceRequirement.getResource()),
+                TextFormat.shortDebugString(mesosResource.getResource()));
+
+        validateRequirements(offerRequirement);
+        setProtos(offerRequirement, fulfilledResource);
+    }
+
+    protected void validateRequirements(OfferRequirement offerRequirement) throws OfferEvaluationException {
+        if (!getTaskName().isPresent() && offerRequirement.getExecutorRequirementOptional().isPresent()) {
```

Comment:
Yeah, further documentation is probably warranted -- the idea is that evaluation stages have a task affinity, since e.g. a dynamic port may need to modify the task that it belongs to, so that's what `taskName` does. If it's not present, then the evaluation stage instead has executor affinity and modifies the `ExecutorInfo` instead, although that functionality is dormant right now since there's not an `ExecutorSpecification`.